core/src/main/scala/pureconfig/ProductHint.scala

```diff
@@ -0,0 +1,40 @@
+package pureconfig
+
+/**
+ * A trait that can be implemented to customize how case classes are read from and written to a config.
+ *
+ * @tparam T the type of case class for which this hint applies
+ */
+trait ProductHint[T] {
+
+  /**
+   * The mapping used to map field names to config keys
+   */
+  def fieldMapping: ConfigFieldMapping
```

Comment:
I agree that without its the implicit usage, `ConfigFieldMapping` is not much more than a `String => String`. Having a way to create such a function object - at any time, and not only via a `ProductHint` - still seems useful, though. First, you can use an independent `ConfigFieldMapping` to manually convert fields when you are implementing a custom `ConfigConvert`. Second, it is slightly more elegant to pass a function to the `ProductHint` constructor than passing two `NamingConvention`s without extra context.

How about we leave the `ConfigFieldMapping` object and make its `apply` methods simply return `String => String`? We could then accept any `String => String` function in hints and use the interface you proposed. The downside is less context - when a `ProductHint` receives a string function, how is it applied? Is it a function from config keys to fields or viceversa?